### ISSUE TITLE
Random seed is not working

### ISSUE THREAD

**Issue by amirmohammadkz** [MilaNLProc/contextualized-topic-models#106]
* Contextualized Topic Models version: 2.2.1
* Python version: 3.9
* Operating System: Windows 10

### Description

I am trying to use CTM, but the results I get during several runs with the same parameters are different. So, the topics I get are not reproducible. 

### What I Did

Before creating the model, I specified all random seeds as it was done in one of the tutorials.
```
    torch.manual_seed(10)
    torch.cuda.manual_seed(10)
    np.random.seed(10)
    random.seed(10)
    nltk.download('stopwords')
    torch.backends.cudnn.enabled = False
    torch.backends.cudnn.deterministic = True

```

However, still I get different topics with different runs.

**Comment by silviatti** [comment]
Hello,
you have to run again the lines to fix the seed (the ones that you reported above) and then run the topic model for the second time. 

Let me know if you still have problems with this.

**Comment by amirmohammadkz** [comment]
Hello,
Yes, the function I am using fixes the seed each before running the TM. The problem is while I am fixing the seed with the same seed number, I get different results.

**Comment by silviatti** [comment]
Can you please send me an extract of your code? I tried with one of our notebooks and it worked for me. Thank you,

Silvia

**Comment by amirmohammadkz** [comment]
Here you are. For `df`, please load a dataframe containing at least a column with text.
```
import random
import numpy as np
import torch
from contextualized_topic_models.models.ctm import CombinedTM
from contextualized_topic_models.models.kitty_classifier import Kitty
import nltk
from contextualized_topic_models.utils.data_preparation import TopicModelDataPreparation
from contextualized_topic_models.utils.preprocessing import WhiteSpacePreprocessing
from nltk.corpus import stopwords
from psylap.utils.data_loader import load_batch, load_data
import string
import unicodedata, re, sys
from tqdm import tqdm
import os
import pandas as pd
from gensim.models.word2vec import Word2Vec
import gensim.downloader as api
from scipy.spatial.distance import cosine
import numpy as np
from nltk.corpus import stopwords as stop_words
from sklearn.feature_extraction.text import CountVectorizer
import warnings
from gensim.utils import deaccent


class MyWhiteSpacePreprocessingStopwords():
    """
    Provides a very simple preprocessing script that filters infrequent tokens from text
    """

    def __init__(self, documents, stopwords_list=None, vocabulary_size=2000, max_df=1.0, min_words=1,
                 remove_numbers=True):
        """

        :param documents: list of strings
        :param stopwords_list: list of the stopwords to remove
        :param vocabulary_size: the number of most frequent words to include in the documents. Infrequent words will be discarded from the list of preprocessed documents
        :param max_df : float or int, default=1.0
        When building the vocabulary ignore terms that have a document
        frequency strictly higher than the given threshold (corpus-specific
        stop words).
        If float in range [0.0, 1.0], the parameter represents a proportion of
        documents, integer absolute counts.
        This parameter is ignored if vocabulary is not None.
        :param min_words: int, default=1. Documents with less words than the parameter
        will be removed
        :param remove_numbers: bool, default=True. If true, numbers are removed from docs
        """
        self.documents = documents
        if stopwords_list is not None:
            self.stopwords = set(stopwords_list)
        else:
            self.stopwords = []

        self.vocabulary_size = vocabulary_size
        self.max_df = max_df
        self.min_words = min_words
        self.remove_numbers = remove_numbers

    def preprocess(self):
        """
        Note that if after filtering some documents do not contain words we remove them. That is why we return also the
        list of unpreprocessed documents.

        :return: preprocessed documents, unpreprocessed documents and the vocabulary list
        """
        preprocessed_docs_tmp = self.documents
        preprocessed_docs_tmp = [deaccent(doc.lower()) for doc in preprocessed_docs_tmp]
        preprocessed_docs_tmp = [doc.translate(
            str.maketrans(string.punctuation, ' ' * len(string.punctuation))) for doc in preprocessed_docs_tmp]
        if self.remove_numbers:
            preprocessed_docs_tmp = [doc.translate(str.maketrans("0123456789", ' ' * len("0123456789")))
                                     for doc in preprocessed_docs_tmp]
        preprocessed_docs_tmp = [' '.join([w for w in doc.split() if len(w) > 0 and w not in self.stopwords])
                                 for doc in preprocessed_docs_tmp]

        vectorizer = CountVectorizer(max_features=self.vocabulary_size, max_df=self.max_df)
        vectorizer.fit_transform(preprocessed_docs_tmp)
        temp_vocabulary = set(vectorizer.get_feature_names())

        preprocessed_docs_tmp = [' '.join([w for w in doc.split() if w in temp_vocabulary])
                                 for doc in preprocessed_docs_tmp]

        preprocessed_docs, unpreprocessed_docs, is_removed = [], [], []
        for i, doc in enumerate(preprocessed_docs_tmp):
            if len(doc) > 0 and len(doc) >= self.min_words:
                preprocessed_docs.append(doc)
                unpreprocessed_docs.append(self.documents[i])
                is_removed.append(0)
            else:
                is_removed.append(1)
        vocabulary = list(set([item for doc in preprocessed_docs for item in doc.split()]))

        return preprocessed_docs, unpreprocessed_docs, vocabulary, is_removed


class TopicModel:
    def __init__(self):
        self.model = "paraphrase-distilroberta-base-v2"
        self.topics = 5
        self.epochs = 10
        self.contextual_size = 768
        self.n_words = 2000
        self.ctm = None
        self.stopwords = set(stop_words.words('english'))
        self.training_dataset = None
        self.tp = None
        self.is_removed = None

    def set_model(self, model):
        self.model = model

    def set_topics(self, topics):
        self.topics = topics

    def set_epochs(self, epochs):
        self.epochs = epochs

    def set_stopwords(self, stop_words_language):
        self.stopwords = set(stop_words.words(stop_words_language))

    def set_contextual_size(self, contextual_size):
        self.contextual_size = contextual_size

    def set_n_words(self, n_words):
        self.n_words = n_words

    def get_word_classes(self, nums=10):
        return self.ctm.get_topic_lists(nums)

    def get_model(self):
        return self.ctm

    def get_lda_data_format(self):
        return self.ctm.get_ldavis_data_format(self.tp.vocab, self.training_dataset, n_samples=10)

    def prepare_dataset(self, df, txt_col):
        nltk.download('stopwords')
        df = df.copy()
        df['training'] = df[txt_col].progress_apply(lambda x: remove_unprintable_chars(x) if pd.notnull(x) else '')

        documents = df['training'].tolist()
        sp = MyWhiteSpacePreprocessingStopwords(documents, stopwords_list=self.stopwords)
        preprocessed_documents, unpreprocessed_corpus, vocab, self.is_removed = sp.preprocess()

        self.tp = TopicModelDataPreparation(self.model)

        self.training_dataset = self.tp.fit(text_for_contextual=unpreprocessed_corpus,
                                            text_for_bow=preprocessed_documents)

    def load_tm(self, dir, epoch):
        self.ctm = CombinedTM(bow_size=len(self.tp.vocab), contextual_size=self.contextual_size,
                              n_components=self.topics,
                              num_epochs=self.epochs, batch_size=128)
        self.ctm.load(dir, epoch)

    def extract_topic_modelling(self):
        self.ctm = CombinedTM(bow_size=len(self.tp.vocab), contextual_size=self.contextual_size,
                              n_components=self.topics,
                              num_epochs=self.epochs, batch_size=128)
        self.ctm.fit(self.training_dataset)  # run the model

        print(self.ctm.get_topic_lists(10))

    def predict_topic_models(self):
        preds = self.ctm.get_thetas(self.training_dataset)
        final_preds = np.zeros((len(self.is_removed), preds.shape[1]))
        cnt = 0
        for i in range(len(self.is_removed)):
            if self.is_removed[i]:
                final_preds[i] = 1 / len(final_preds[i])
            else:
                final_preds[i] = preds[cnt]
                cnt += 1
        res_df = pd.DataFrame(final_preds)
        word_classes = self.get_word_classes(5)
        word_classes = ['_'.join(word_class) for word_class in word_classes]
        res_df.columns = word_classes
        res_df['is_removed'] = self.is_removed
        return res_df


def compute_topic_coherence(word_classes, wv=None):
    def word_list2vecs(wv, word_list):
        vecs = []
        for word in word_list:
            try:
                vecs.append(wv[word])
            except KeyError:
                continue
        return vecs

    def compute_external_coherence(vec_list):
        cos_list = []
        for i in range(len(vec_list)):
            for j in range(i, len(vec_list)):
                if not i == j:
                    cos_list.append(cosine(vec_list[i], vec_list[j]))
        return np.mean(cos_list)

    if wv is None:
        wv = api.load('word2vec-google-news-300')

    vec_classes = []
    for word_class in word_classes:
        vec_classes.append(word_list2vecs(wv, word_class))

    cos_list = []
    for vec_class in vec_classes:
        cos_list.append(compute_external_coherence(vec_class))

    return np.mean(cos_list)


def extract_and_predict_optimum_tms(df, txt_col, id_col, model='all-mpnet-base-v2', topics=None):
    if topics is None:
        topics = [3, 5, 7, 9, 11]
    torch.manual_seed(10)
    torch.cuda.manual_seed(10)
    np.random.seed(10)
    random.seed(10)
    nltk.download('stopwords')
    torch.backends.cudnn.enabled = False
    torch.backends.cudnn.deterministic = True
    tqdm.pandas()
    print("Loading google w2v...")
    wv = api.load('word2vec-google-news-300')
    cos_scores = []
    tms = []
    for topic_num in topics:
        print('TM with {} topics...'.format(topic_num))
        tm = TopicModel()
        tm.set_model(model)
        tm.set_topics(topic_num)
        tm.prepare_dataset(df, txt_col)
        tm.extract_topic_modelling()
        print('Calculating external word embeddings topic coherence...')
        cos_scores.append(compute_topic_coherence(tm.get_word_classes(), wv))
        print("Score: {}".format(cos_scores[-1]))
        tms.append(tm)
    tm = tms[np.argmin(cos_scores)]
    res = tm.predict_topic_models()
    lda_data = tm.get_lda_data_format()
    ctm = tm.get_model()

    return res, lda_data, ctm


def load_and_predict(df, txt_col, tm_dir, epoch, topic_num, model='all-mpnet-base-v2'):
    torch.manual_seed(10)
    torch.cuda.manual_seed(10)
    np.random.seed(10)
    random.seed(10)
    nltk.download('stopwords')
    torch.backends.cudnn.enabled = False
    torch.backends.cudnn.deterministic = True
    tqdm.pandas()

    tm = TopicModel()
    tm.set_model(model)
    tm.set_topics(topic_num)
    tm.prepare_dataset(df, txt_col)
    tm.load_tm(tm_dir, epoch)

    res = tm.predict_topic_models()
    lda_data = tm.get_lda_data_format()
    ctm = tm.get_model()

    return res, lda_data, ctm


def remove_unprintable_chars(s):
    return ''.join([i for i in s if i in string.printable])


if __name__ == "__main__":

    df = load_data("../../../resources/complete/data/MyEveningJournal.csv")
    res, lda_data, ctm = extract_and_predict_optimum_tms(df, 'title', 'participantId', model='all-mpnet-base-v2', topics=[4, 5])
    print("done")


```

**Comment by vinid** [comment]
see this [notebook](https://colab.research.google.com/drive/1KO6SH1Fg5ituSaw0dzENiTLcgj1yNSYG?usp=sharing)

You will see that the predictions are always the same, even after saving and reloading the model

**Comment by amirmohammadkz** [comment]
Hi @vinid,
I made an edit and it does not work anymore:
https://colab.research.google.com/drive/1BkfI2XZwPenxFenlGc5LAj4TPunnXEwY?usp=sharing

Can you please guide me to find the problem?

**Comment by vinid** [comment]
[here it is](https://colab.research.google.com/drive/1wN3cIRX6TP6KGKJbirvj_tjN6tcFZ0hd?usp=sharing). 

If you want a stochastic operation to always have the same result you need to reset the seed every time before you call it.

```python

import random

random.seed(4)
random.random()
>>> 0.23604808973743452

random.random()
>>> 0.1031660342307158

random.seed(4)
random.random()
>>> 0.23604808973743452
```

**Comment by vinid** [comment]
sure! let me know how it goes :)
